lib/src/equals_matcher.dart

```diff
@@ -260,16 +269,48 @@ class _DeepMatcher extends Matcher {
   @override
   Description describeMismatch(
       item, Description mismatchDescription, Map matchState, bool verbose) {
-    var reason = matchState['reason'] as String ?? '';
-    // If we didn't get a good reason, that would normally be a
-    // simple 'is <value>' message. We only add that if the mismatch
-    // description is non empty (so we are supplementing the mismatch
-    // description).
-    if (reason.isEmpty && mismatchDescription.length > 0) {
-      mismatchDescription.add('is ').addDescriptionOf(item);
+    var mismatch = matchState['mismatch'] as _Mismatch;
+    if (mismatch.location.isNotEmpty) {
+      mismatchDescription
+          .add('at location ')
+          .add(mismatch.location)
+          .add(' is ')
+          .addDescriptionOf(mismatch.actual);
+      if (mismatch.describeProblem != null) {
+        mismatchDescription
+            .add(' ${mismatch.instead ? 'instead of ' : 'which'} ');
+        mismatch.describeProblem(mismatchDescription, verbose);
+      }
     } else {
-      mismatchDescription.add(reason);
+      // If we didn't get a good reason, that would normally be a
+      // simple 'is <value>' message. We only add that if the mismatch
+      // description is non empty (so we are supplementing the mismatch
+      // description).
+      if (mismatch.describeProblem == null) {
+        if (mismatchDescription.length > 0) {
+          mismatchDescription.add('is ').addDescriptionOf(item);
+        }
+      } else {
+        mismatch.describeProblem(mismatchDescription, verbose);
+      }
     }
     return mismatchDescription;
   }
 }
+
+class _Mismatch {
+  final String location;
+
+  final Object actual;
+
+  final void Function(Description description, bool verbose) describeProblem;
```

Comment:
[nit] We don't have to specify a variable name in a function type, and where it matches the type I prefer to omit it.

```suggestion
  final void Function(Description, bool verbose) describeProblem;
```

Reply:
Done